core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ast/SparqlFragment.java

```diff
@@ -7,29 +7,175 @@
 
+import static org.eclipse.rdf4j.sail.shacl.ast.constraintcomponents.AbstractConstraintComponent.VALUES_INJECTION_POINT;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
 public class SparqlFragment {
 
-	String fragment;
-	boolean filterCondition;
-	boolean bgp;
+	private final String fragment;
+	private final List<String> unionFragments = new ArrayList<>();
+
+	private boolean filterCondition;
+	private boolean bgp;
+	private boolean union;
+
+	// This is currently experimental!
+	private static final boolean USE_UNION_PRESERVING_JOIN = false;
+
+	private final List<StatementMatcher> statementMatchers = new ArrayList<>();
+
+	public SparqlFragment(List<String> unionFragments) {
+		this.fragment = null;
+		this.unionFragments.addAll(unionFragments);
+		this.union = true;
+	}
 
 	public SparqlFragment(String fragment) {
 		this.fragment = fragment;
 	}
 
-	public static SparqlFragment filterCondition(String fragment) {
+	public SparqlFragment(String fragment, List<StatementMatcher> statementMatchers) {
+		this.fragment = fragment;
+		this.statementMatchers.addAll(statementMatchers);
+	}
+
+	public static SparqlFragment filterCondition(String fragment, List<StatementMatcher> statementMatchers) {
 		SparqlFragment sparqlFragment = new SparqlFragment(fragment);
 		sparqlFragment.filterCondition = true;
+		sparqlFragment.statementMatchers.addAll(statementMatchers);
 		return sparqlFragment;
 	}
 
-	public static SparqlFragment bgp(String fragment) {
+	public static SparqlFragment bgp(String fragment, List<StatementMatcher> statementMatchers) {
 		SparqlFragment sparqlFragment = new SparqlFragment(fragment);
 		sparqlFragment.bgp = true;
+		sparqlFragment.statementMatchers.addAll(statementMatchers);
+		return sparqlFragment;
+	}
+
+	public static SparqlFragment and(List<SparqlFragment> sparqlFragments) {
+		String collect = sparqlFragments.stream()
+				.peek(s -> {
+					assert s.filterCondition;
+				})
+				.map(SparqlFragment::getFragment)
+				.collect(Collectors.joining(" ) && ( ", "( ",
+						" )"));
+
+		return filterCondition(collect,
+				getStatementMatchers(sparqlFragments));
+	}
+
+	public static SparqlFragment or(List<SparqlFragment> sparqlFragments) {
+		String collect = sparqlFragments.stream()
+				.peek(s -> {
+					assert s.filterCondition;
+				})
+				.map(SparqlFragment::getFragment)
+				.collect(Collectors.joining(" ) || ( ", "( ",
+						" )"));
+
+		return filterCondition(collect,
+				getStatementMatchers(sparqlFragments));
+	}
+
+	public static SparqlFragment join(List<SparqlFragment> sparqlFragments) {
+
+		if (USE_UNION_PRESERVING_JOIN && sparqlFragments.stream().anyMatch(s1 -> s1.union)) {
+			return unionPreservingJoin(sparqlFragments);
+
+		} else {
+			String collect = sparqlFragments.stream()
+					.peek(s -> {
+						assert !s.filterCondition;
+					})
+					.map(SparqlFragment::getFragment)
+					.map(s -> s.replaceAll("(?m)^", "\t"))
```

Comment:
If this is going to stay long term, maybe precompile the pattern

Reply:
It's to make the query a bit more readable when debugging. I'll compile it, that'll help a lot! Thanks.

Reply:
Btw. The thing that is currently using the most time is that the generated SPARQL query needs to be optimized and evaluated. That's about 99% of the execution time.